text/0000-structural-records.md

```diff
@@ -0,0 +1,1208 @@
+*"unnamed structs"*, or *"anonymous structs"*.
+
+# Motivation
+[motivation]: #motivation
+
+There are four motivations to introduce structural records,
+two major and two minor.
+
+## Major: Improving ergonomics, readability, and maintainability
+
+Sometimes, you just need a one-off struct that you will use a few times for
+some public API or in particular for some internal API. For example, you'd like
+to send over a few values to a function, but you don't want to have too many
+function arguments. In such a scenario, defining a new nominal `struct` which
+contains all the fields you want to pass is not a particularly ergonomic solution.
+Instead, it is rather heavy weight:
+
+```rust
+struct Color {
+    red: u8,
+    green: u8,
+    blue: u8,
+}
+
+fn caller() {
+    ...
+    let color = Color { red: 255, green: 0, blue: 0 };
+    ...
+    do_stuff_with(color);
+    ...
+}
+
+fn do_stuff_with(color: Color) { // Only time we use `Color`!
+    some_stuff(color.red);
+    ...
+    other_stuff(color.green);
+    ...
+    yet_more_stuff(color.blue);
+}
+```
+
+To remedy the ergonomics problem, you may instead opt for using a positional
+tuple to contain the values you want to pass. However, now you have likely
+created a regression for readers of your code since the fields of tuples are
+accessed with their positional index, which does not carry clear semantic intent:
+
+```rust
+fn caller() {
+    ...
+    let color = (255, 0, 0);
+    ...
+    do_stuff_with(color); // Unclear what each position means...
+    ...
+}
+
+fn do_stuff_with(color: (u8, u8, u8)) { // More ergonomic!
+    some_stuff(color.0); // But less readable... :(
+    ...
+    other_stuff(color.1);
+    ...
+    yet_more_stuff(color.2);
+}
+```
+
+Using structural records, we can have our cake and eat it too:
+
+```rust
+fn caller() {
+    ...
+    let color = { red: 255, green: 0, blue: 0 };
+    ...
+    do_stuff_with(color); // ...but here it is clear.
+    ...
+}
+
+fn do_stuff_with(color: { red: u8, green: u8, blue: u8 }) { // More ergonomic!
+    some_stuff(color.red); // *And* readable! :)
+    ...
+    other_stuff(color.green);
+    ...
+    yet_more_stuff(color.blue);
+}
+```
+
+In the above snippet, the semantic intent of the fields is clear both when
+reading the body of the function, as well as when reading the documentation
+when `do_stuff_with` is exposed as a public API.
+
+[@eternaleye]: https://internals.rust-lang.org/t/pre-rfc-unnamed-struct-types/3872/58
+[@kardeiz]: https://internals.rust-lang.org/t/pre-rfc-unnamed-struct-types/3872/65
+
+Another example of reducing boilerplate was given by [@eternaleye]:
+
+```rust
+struct RectangleClassic {
+    width: u64,
+    height u64,
+    red: u8,
+    green: u8,
+    blue: u8,
+}
+
+struct RectangleTidy {
+    dimensions: {
+        width: u64,
+        height: u64,
+    },
+    color: {
+        red: u8,
+        green: u8,
+        blue: u8,
+    },
+}
+```
```

Comment:
This syntax looks very similar to another already accepted RFC, but semantics seems different - https://github.com/rust-lang/rfcs/pull/2102.